cloud/pkg/edgecontroller/controller/downstream.go

```diff
@@ -425,6 +438,118 @@ func (dc *DownstreamController) syncEndpoints() {
+		return nil
+	}
+	bkd := &api.Backend{
+		Name:            svc.Name,
+		Namespace:       svc.Namespace,
+		ResourceVersion: svc.ResourceVersion,
+		HostNetwork:     pods[podIndex].Spec.HostNetwork,
+		Subsets:         make([]api.BackendSubset, 0),
+		Ports:           make([]api.BackendPort, 0),
+	}
+	for _, pod := range pods {
+		bs := api.BackendSubset{
+			PodIP:    pod.Status.PodIP,
+			HostIP:   pod.Status.HostIP,
+			State:    string(pod.Status.Phase),
+			NodeName: pod.Spec.NodeName,
+		}
+		if pod.DeletionTimestamp == nil {
+			bkd.Subsets = append(bkd.Subsets, bs)
+		}
+	}
+	hostAddrMap := make(map[int32][]string) // handle case that multiple pods of one deployment on the same edge node
+	for _, pod := range pods {
+		if pod.Spec.HostNetwork {
+			continue
+		}
+		hostIP := pod.Status.HostIP
+		for _, container := range pod.Spec.Containers {
+			for _, port := range container.Ports {
+				hostAddr := hostIP + AddrSeparator + strconv.FormatInt(int64(port.HostPort), 10)
+				addr, ok := hostAddrMap[port.ContainerPort]
+				if ok {
+					addr = append(addr, hostAddr)
+					hostAddrMap[port.ContainerPort] = addr
+					continue
+				}
+				hostAddrMap[port.ContainerPort] = []string{hostAddr}
+			}
+		}
+	}
+	for _, port := range svc.Spec.Ports {
+		bp := api.BackendPort{
+			Name:          port.Name,
+			ContainerPort: port.TargetPort.IntVal,
+			HostPort:      hostAddrMap[port.TargetPort.IntVal],
+		}
+		bkd.Ports = append(bkd.Ports, bp)
+	}
+	return bkd
+}
+
+// get service backend by service endpoints and service pod list, send backend instead of pod list to edge
+func getBackendByEndpointsAndPods(eps *v1.Endpoints, pods []v1.Pod) *api.Backend {
+	if len(pods) == 0 {
+		klog.Warningf("endpoints name: %s, namespace: %s with no pods", eps.Name, eps.Namespace)
+		return nil
+	}
+	bkd := &api.Backend{
+		Name:            eps.Name,
+		Namespace:       eps.Namespace,
+		ResourceVersion: eps.ResourceVersion,
+		HostNetwork:     pods[podIndex].Spec.HostNetwork,
+		Subsets:         make([]api.BackendSubset, 0),
+		Ports:           make([]api.BackendPort, 0),
+	}
+	for _, pod := range pods {
+		bs := api.BackendSubset{
+			PodIP:    pod.Status.PodIP,
+			HostIP:   pod.Status.HostIP,
+			State:    string(pod.Status.Phase),
+			NodeName: pod.Spec.NodeName,
+		}
+		if pod.DeletionTimestamp == nil {
+			bkd.Subsets = append(bkd.Subsets, bs)
+		}
+	}
+	hostAddrMap := make(map[int32][]string) // handle case that multiple pods of one deployment on the same edge node
+	for _, pod := range pods {
+		if pod.Spec.HostNetwork {
+			continue
+		}
+		hostIP := pod.Status.HostIP
+		for _, container := range pod.Spec.Containers {
+			for _, port := range container.Ports {
+				hostAddr := hostIP + AddrSeparator + strconv.FormatInt(int64(port.HostPort), 10)
+				addr, ok := hostAddrMap[port.ContainerPort]
+				if ok {
+					addr = append(addr, hostAddr)
+					hostAddrMap[port.ContainerPort] = addr
```

Comment:
Users must specify HostPort and ContainerPort for the Container when they want to use edgemesh? Can we get this value from service and endpoint.

Reply:
It seems that in the current use case it must specify HostPort and ContainerPort for the Container , ignore this comment.